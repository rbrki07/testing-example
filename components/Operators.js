// @ts-check
import typedefs from "./../typedefs";
import React from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";

/**
 * @param {Object} params
 * @param {typedefs.Operator} params.operator
 *
 * @returns {string}
 */
const renderOperator = ({ operator }) => {
  switch (operator) {
    case "add":
      return "+";
    case "substract":
      return "-";
    case "multiply":
      return "*";
    case "divide":
      return "/";
  }
};

/**
 * @param {Object} params
 * @param {typedefs.Operator} params.operator
 * @param {boolean} params.isSelected
 * @param {(operator: typedefs.OperatorObject) => void} params.setOperator
 *
 * @returns {Object}
 */
const Operator = ({ operator, isSelected, setOperator }) => {
  return (
    <TouchableOpacity
      style={[
        styles.operator,
        isSelected
          ? { backgroundColor: "#f1faee" }
          : { backgroundColor: "#fff" },
      ]}
      onPress={() => setOperator({ operator })}
    >
      <Text>{renderOperator({ operator })}</Text>
    </TouchableOpacity>
  );
};

/**
 * @param {Object} params
 * @param {typedefs.Operator} params.operator
 * @param {(operator: typedefs.OperatorObject) => void} params.setOperator
 *
 * @returns {Object}
 */
const Operators = ({ operator, setOperator }) => {
  return (
    <View style={styles.container}>
      <Operator
        operator={"add"}
        isSelected={"add" === operator}
        setOperator={setOperator}
      />
      <Operator
        operator={"substract"}
        isSelected={"substract" === operator}
        setOperator={setOperator}
      />
      <Operator
        operator={"multiply"}
        isSelected={"multiply" === operator}
        setOperator={setOperator}
      />
      <Operator
        operator={"divide"}
        isSelected={"divide" === operator}
        setOperator={setOperator}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 48,
    flexDirection: "row",
  },
  operator: {
    height: 32,
    width: 32,
    alignItems: "center",
    justifyContent: "center",
    margin: 4,
    paddingLeft: 4,
    paddingRight: 4,
    borderRadius: 8,
    fontSize: 22,
    shadowOffset: {
      width: 1,
      height: 1,
    },
    shadowOpacity: 0.75,
    shadowRadius: 1,
  },
});

export { Operators };
