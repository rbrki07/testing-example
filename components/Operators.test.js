// @ts-check
import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { Operators } from "./Operators";

describe("Operators component", () => {
  it("should set `isSelected` correctly", () => {
    const setOperator = jest.fn();
    let operator = "add";

    const { getByText } = render(
      // @ts-ignore
      <Operators operator={operator} setOperator={setOperator} />
    );

    let component = getByText("+");
    expect(component.parent?.props.style).toMatchObject({
      backgroundColor: "#f1faee",
    });

    fireEvent.press(component);
    expect(setOperator).toHaveBeenCalledWith({ operator });

    component = getByText("-");
    expect(component.parent?.props.style).toMatchObject({
      backgroundColor: "#fff",
    });
  });

  it("should render correctly", () => {
    const setOperator = jest.fn();
    const operator = "add";
    expect(
      render(
        <Operators operator={operator} setOperator={setOperator} />
      ).toJSON()
    ).toMatchSnapshot();
  });
});
