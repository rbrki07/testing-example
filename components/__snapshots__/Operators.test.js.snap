// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operators component should render correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "height": 48,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f1faee",
        "borderRadius": 8,
        "fontSize": 22,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 4,
        "paddingRight": 4,
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 1,
        "width": 32,
      }
    }
  >
    <Text>
      +
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "fontSize": 22,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 4,
        "paddingRight": 4,
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 1,
        "width": 32,
      }
    }
  >
    <Text>
      -
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "fontSize": 22,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 4,
        "paddingRight": 4,
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 1,
        "width": 32,
      }
    }
  >
    <Text>
      *
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "fontSize": 22,
        "height": 32,
        "justifyContent": "center",
        "margin": 4,
        "opacity": 1,
        "paddingLeft": 4,
        "paddingRight": 4,
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 0.75,
        "shadowRadius": 1,
        "width": 32,
      }
    }
  >
    <Text>
      /
    </Text>
  </View>
</View>
`;
